pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java

```diff
@@ -133,183 +122,144 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
+        // Initialize Null value vector map
+        String columnName = fieldSpec.getName();
+        _nullValueVectorCreatorMap.put(columnName, new NullValueVectorCreator(_indexDir, columnName));
       }
+    }
+  }
 
-      JsonIndexConfig jsonIndexConfig = jsonIndexConfigs.get(columnName);
-      if (jsonIndexConfig != null) {
-        _jsonIndexCreatorMap.put(columnName,
-            _indexCreatorProvider.newJsonIndexCreator(context.forJsonIndex(jsonIndexConfig)));
-      }
+  private FieldIndexConfigs adaptConfig(String columnName, FieldIndexConfigs config,
+      ColumnIndexCreationInfo columnIndexCreationInfo, SegmentGeneratorConfig segmentCreationSpec) {
+    FieldIndexConfigs.Builder builder = new FieldIndexConfigs.Builder(config);
+    // Sorted columns treat the 'forwardIndexDisabled' flag as a no-op
+    if (config.getConfig(StandardIndexes.forward()).isEnabled() && columnIndexCreationInfo.isSorted()) {
+      builder.add(StandardIndexes.forward(), new ForwardIndexConfig.Builder()
+          .withLegacyProperties(segmentCreationSpec.getColumnProperties(), columnName)
+          .build());
+    }
+    // Initialize inverted index creator; skip creating inverted index if sorted
+    if (config.getConfig(StandardIndexes.inverted()).isEnabled() && columnIndexCreationInfo.isSorted()) {
+      builder.undeclare(StandardIndexes.inverted());
+    }
+    return builder.build();
+  }
 
-      H3IndexConfig h3IndexConfig = h3IndexConfigs.get(columnName);
-      if (h3IndexConfig != null) {
-        _h3IndexCreatorMap.put(columnName,
-            _indexCreatorProvider.newGeoSpatialIndexCreator(context.forGeospatialIndex(h3IndexConfig)));
-      }
+  /**
+   * Returns true if the given index type has their own construction lifecycle and therefore should not be instantiated
+   * in the general index loop and shouldn't be notified of each new column.
+   */
+  private boolean skipIndexType(IndexType<?, ?, ?> indexType) {
+    return indexType == StandardIndexes.nullValueVector() || indexType == StandardIndexes.dictionary();
+  }
 
-      _nullHandlingEnabled = _config.isNullHandlingEnabled();
-      if (_nullHandlingEnabled) {
-        // Initialize Null value vector map
-        _nullValueVectorCreatorMap.put(columnName, new NullValueVectorCreator(_indexDir, columnName));
-      }
+  /**
+   * Creates the {@link IndexCreator} in a type safe way.
+   *
+   * This code needs to be in a specific method instead of inlined in the main loop in order to be able to use the
+   * limited generic capabilities of Java.
+   */
+  private <C extends IndexConfig> void tryCreateCreator(Map<IndexType<?, ?, ?>, IndexCreator> creatorsByIndex,
+      IndexType<C, ?, ?> index, IndexCreationContext.Common context, FieldIndexConfigs fieldIndexConfigs)
+      throws Exception {
+    C declaration = fieldIndexConfigs.getConfig(index);
+    if (declaration.isEnabled() && index.shouldBeCreated(context, fieldIndexConfigs)) {
```

Comment:
s/declaration/config? 

I see IndexDeclaration was mentioned as an alternative in the PEP but looks like this PR is not using that way?